// @aws-lambda-powertools/metrics/lib/types/MetricUnit.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MetricUnits = void 0;
var MetricUnits;
(function (MetricUnits) {
    MetricUnits["Seconds"] = "Seconds";
    MetricUnits["Microseconds"] = "Microseconds";
    MetricUnits["Milliseconds"] = "Milliseconds";
    MetricUnits["Bytes"] = "Bytes";
    MetricUnits["Kilobytes"] = "Kilobytes";
    MetricUnits["Megabytes"] = "Megabytes";
    MetricUnits["Gigabytes"] = "Gigabytes";
    MetricUnits["Terabytes"] = "Terabytes";
    MetricUnits["Bits"] = "Bits";
    MetricUnits["Kilobits"] = "Kilobits";
    MetricUnits["Megabits"] = "Megabits";
    MetricUnits["Gigabits"] = "Gigabits";
    MetricUnits["Terabits"] = "Terabits";
    MetricUnits["Percent"] = "Percent";
    MetricUnits["Count"] = "Count";
    MetricUnits["BytesPerSecond"] = "Bytes/Second";
    MetricUnits["KilobytesPerSecond"] = "Kilobytes/Second";
    MetricUnits["MegabytesPerSecond"] = "Megabytes/Second";
    MetricUnits["GigabytesPerSecond"] = "Gigabytes/Second";
    MetricUnits["TerabytesPerSecond"] = "Terabytes/Second";
    MetricUnits["BitsPerSecond"] = "Bits/Second";
    MetricUnits["KilobitsPerSecond"] = "Kilobits/Second";
    MetricUnits["MegabitsPerSecond"] = "Megabits/Second";
    MetricUnits["GigabitsPerSecond"] = "Gigabits/Second";
    MetricUnits["TerabitsPerSecond"] = "Terabits/Second";
    MetricUnits["CountPerSecond"] = "Count/Second";
})(MetricUnits || (MetricUnits = {}));
exports.MetricUnits = MetricUnits;
